cs/src/core/RuntimeDeserializer.cs

```diff
@@ -0,0 +1,671 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Bond
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using Bond.IO.Safe;
+    using Bond.Protocols;
+
+    /// <summary>
+    /// Deserializes <see cref="RuntimeObject"/>s from a protocol reader using only a runtime schema.
+    /// </summary>
+    public static class RuntimeDeserialize
+    {
+        /// <summary>
+        /// Deserializes a <see cref="RuntimeObject"/> from a tagged protocol reader using only a runtime schema.
+        /// </summary>
+        /// <param name="reader">The tagged protocol reader to read the object data from.</param>
+        /// <param name="schema">The schema to use when reading the object data.</param>
+        /// <returns>A <see cref="RuntimeObject"/> created from <paramref name="reader"/> and <paramref name="schema"/>.</returns>
+        public static RuntimeObject From(IClonableTaggedProtocolReader reader, RuntimeSchema schema)
+        {
+            return new RuntimeDeserializer(schema).Deserialize(reader);
+        }
+
+        /// <summary>
+        /// Deserializes a <see cref="RuntimeObject"/> from an untagged protocol reader using only a runtime schema.
+        /// </summary>
+        /// <param name="reader">The untagged protocol reader to read the object data from.</param>
+        /// <param name="schema">The schema to use when reading the object data.</param>
+        /// <returns>A <see cref="RuntimeObject"/> created from <paramref name="reader"/> and <paramref name="schema"/>.</returns>
+        public static RuntimeObject From(IClonableUntaggedProtocolReader reader, RuntimeSchema schema)
+        {
+            return new RuntimeDeserializer(schema).Deserialize(reader);
+        }
+    }
+
+    /// <summary>
+    /// Deserializes <see cref="RuntimeObject"/>s from a protocol reader using only a <see cref="RuntimeSchema"/>.
+    /// </summary>
+    public class RuntimeDeserializer
```

Comment:
The fact that existing deserializer is instantiated per C# type is a choice, not necessity. We do this for every C# type because deserializer *is* different for every type.

There is nothing preventing you from having just one instance per protocol. For example that's how [Transcode](https://github.com/Microsoft/bond/blob/1da75c2ffa91d8a20a22ee2cb9714aef28217872/cs/src/core/Transcoder.cs#L10-L13) API works - there is one instance per a pair of protocols. It still reuses the existing parsers.

BTW, it might be good to have an option to generate different deserializer per `RuntimeSchema`. There is potential to get much better performance this way because you can process the `SchemaDef` at expression generation time rather than runtime.

Support for all protocols would work the same way it works for all other transforms. Note for example that `Deserializer` is fully implemented in `Bond.dll`, it knows nothing about `Bond.JSON.dll` and you can [use it to deserialize](https://github.com/Microsoft/bond/blob/1da75c2ffa91d8a20a22ee2cb9714aef28217872/examples/cs/core/simple_json/program.cs#L28) from JSON.

If you don't want to go full-blown transform-based deserializer, there might be another approach that is much simpler: implement it as `IProtocolWriter` and use transcoding. Since your `RuntimeObject` is simply a dictionary from string to object, and in [WriteFieldBegin](https://github.com/Microsoft/bond/blob/1da75c2ffa91d8a20a22ee2cb9714aef28217872/cs/src/core/protocols/IProtocolWriter.cs#L46) you get the `Metadata` which contains field name, you should be able to implement a state-full writer which puts all the fields into a tree of dictionaries. This way you also avoid duplicating all the parsing code. Have you considered/tried this?